ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,363 @@
+   * <p/>
+   * The expectation is that a matching {@link #closeView()} will be called eventually to decrement the ref count.
+   * @param offset the offset that will used to start the read operation from.
+   * @return the {@link File} and {@link FileChannel} backing this log segment.
+   * @throws IllegalArgumentException if {@code offset} < 0 or is > {@link #getEndOffset()}.
+   */
+  Pair<File, FileChannel> getView(long offset) {
+    if (offset < 0 || offset > getEndOffset()) {
+      throw new IllegalArgumentException("Provided offset [" + offset + "] is out of segment range");
+    }
+    refCount.incrementAndGet();
+    return segmentView;
+  }
+
+  /**
+   * Closes view that was obtained (decrements ref count).
+   */
+  void closeView() {
+    refCount.decrementAndGet();
+  }
+
+  /**
+   * @return size of the backing file on disk.
+   * @throws IOException if the size could not be obtained due to I/O error.
+   */
+  long sizeInBytes()
+      throws IOException {
+    return fileChannel.size();
+  }
+
+  /**
+   * @return the name of this segment.
+   */
+  String getName() {
+    return name;
+  }
+
+  /**
+   * Sets the end offset of this segment. This can be lesser than the actual size of the file and represents the offset
+   * until which data that is readable is stored.
+   * @param endOffset the end offset of this log.
+   * @throws IllegalArgumentException if {@code endOffset} < 0 or {@code endOffset} > the size of the file.
+   * @throws IOException if there is any I/O error.
+   */
+  void setEndOffset(long endOffset)
+      throws IOException {
+    long fileSize = sizeInBytes();
+    if (endOffset < 0 || endOffset > fileSize) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + ": EndOffset [" + endOffset +
+          "] outside the file size [" + fileSize + "]");
+    }
+    fileChannel.position(endOffset);
+    this.endOffset.set(endOffset);
+  }
+
+  /**
+   * @return the end offset of this log segment.
+   */
+  long getEndOffset() {
+    return endOffset.get();
+  }
+
+  /**
+   * @return the reference count of this log segment.
+   */
+  long refCount() {
+    return refCount.get();
+  }
+
+  /**
+   * Flushes the backing file to disk.
+   * @throws IOException if there is an I/O error while flushing.
+   */
+  void flush()
+      throws IOException {
+    fileChannel.force(true);
+  }
+
+  /**
+   * Closes this log segment
+   */
+  void close()
+      throws IOException {
+    fileChannel.close();
+  }
+
+  /**
+   * Checks to make sure that {@link #next} is not {@code null}
+   * @param errorCounter the error counter to increment if {@link #next} is {@code null}.
+   * @throws IllegalStateException if {@link #next} is {@code null}
+   */
+  private void ensureNext(Counter errorCounter) {
+    if (next == null) {
+      errorCounter.inc();
+      throw new IllegalStateException(file.getAbsolutePath() + ": Cannot perform operation beyond total capacity");
```

Comment:
Then as far as the segment (and the higher log abstraction) is concerned, the total capacity has been reached.